### PR TITLE
kvserver: move RaftTruncatedState from ReplicaState

### PR DESCRIPTION
**Pull request by pav-kv**
The `RaftTruncatedState` is in the unreplicated rangeID-local space. It logically belongs to the raft log storage, and is separate from the replicated state machine state. Before this PR, `RaftTruncatedState` was mixed into the `ReplicaState` struct which contains replicated state such as the range descriptor.

After this PR, `Replica` tracks the `RaftTruncatedState` separately from `ReplicaState`:

- `StateLoader.{Save,Load}` methods ignore this field. It is now saved and loaded using the `RaftTruncatedState` accessors.
- `ReplicaState.TruncatedState` is always nil in memory.
- The only place where the `ReplicaState.TruncatedState` remains used is the `ReplicatedEvalResult.State`. This is needed for compatibility with older versions, because the log may still contain historical proposals.
- However, all new proposals evaluate to a newly introduced `ReplicatedEvalResult.RaftTruncatedState` field. This change is version-gated.

The last remaining use of `ReplicaState.TruncatedState` needs to be phased out with a migration, before the field can be fully removed. All historical entries need to be applied/removed, before it is safe.

Part of #97613

### CODE REVIEW

**Comment by pav-kv** [comment]
@tbg PTAL. I split this change into gradual commits:

- First, introduce a new in-memory field in `Replica` to mirror `r.shMu.state.TruncatedState`. This will later move into the "log storage" type.
- Then, switch to using the new field instead of `r.shMu.state.TruncatedState`.
- Then, maintain `r.shMu.state.TruncatedState == nil` instead of mirroring.
- Factor out truncated state loading and saving from the `ReplicaState` state loader.
- Finally, deal with `TruncatedState` also being used in `ReplicatedEvalResult`.

All commits passed `dev test pkg/kv/kvserver`.

File: pkg/kv/kvserver/replica.go [hunk 1]
```diff
@@ -442,6 +442,13 @@ type Replica struct {
 	shMu struct {
 		// The state of the Raft state machine.
 		state kvserverpb.ReplicaState
+		// raftTruncState contains the raft log truncation state, i.e. the ID of the
+		// last entry of the log prefix that has been compacted out from the raft
+		// log storage.
+		//
+		// Invariant: raftTruncState == *state.TruncatedState.
```
Comment (pav-kv) [inline pkg/kv/kvserver/replica.go:449]: self-nit: this invariant was true in the second commit, but turned into `state.TruncatedState == nil` towards the end. Fix the comment.

File: pkg/kv/kvserver/kvserverpb/proposer_kv.go [hunk 2]
```diff
@@ -72,3 +72,31 @@ func (r *ReplicatedEvalResult) IsTrivial() bool {
 	allowlist.State = nil
 	return allowlist.IsZero()
 }
+
+func (r *ReplicatedEvalResult) SetRaftTruncatedState(ts *RaftTruncatedState, v25dot1 bool) {
```
Comment (pav-kv) [inline pkg/kv/kvserver/kvserverpb/proposer_kv.go:76]: Add some comments.

File: pkg/kv/kvserver/replica_app_batch.go [hunk 3]
```diff
@@ -48,7 +48,9 @@ type replicaAppBatch struct {
 	// to copy-on-write. The exception to this is `state.Stats`, for which
 	// backing memory has already been provided and which may thus be
 	// modified directly.
-	state kvserverpb.ReplicaState
+	state      kvserverpb.ReplicaState
+	truncState kvserverpb.RaftTruncatedState
```
Comment (pav-kv) [inline pkg/kv/kvserver/replica_app_batch.go:52]: Add a comment.

File: pkg/kv/kvserver/stateloader/stateloader.go [hunk 4]
```diff
@@ -80,14 +80,6 @@ func (rsl StateLoader) Load(
 	s.Stats = &ms
 	s.RaftClosedTimestamp = as.RaftClosedTimestamp
```
Comment (pav-kv) [inline pkg/kv/kvserver/stateloader/stateloader.go:82]: Can explicitly say `TruncatedState = nil` here, to stress that this field is unused and should be loaded separately.

**Review by tbg — APPROVED** [review]
:lgtm: and thank you for putting in the effort to make this an easy review.

> Reviewed 30 of 30 files at r8, 5 of 5 files at r9, 7 of 7 files at r10, 8 of 8 files at r11, 3 of 3 files at r12, 5 of 5 files at r13, 13 of 13 files at r14, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136156#-:-OCnY0u5-1sfKCyzBOto:br3ku6h)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @pav-kv)

___
*[`pkg/kv/kvserver/kvserverpb/proposer_kv.go` line 94 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/136156#-OCnWO_u0HbtXefS1onS:-OCnWO_u0HbtXefS1onT:b-l2mu2x) ([raw file](https://github.com/cockroachdb/cockroach/blob/873795935fbfe1f5eb3f2d4a1e87038b93d46af4/pkg/kv/kvserver/kvserverpb/proposer_kv.go#L94)):*
> ```Go
> // UnsetRaftTruncatedState clears the RaftTruncatedState from this evaluation
> // result.
> func (r *ReplicatedEvalResult) UnsetRaftTruncatedState() {
> ```

This is doing more than I expected. Does this method make sense? It's called only once in a specific situation, separating this out makes things less clear if anything.

Also do we have to worry about the case in which `r.State` contains only the `TruncatedState`? Do we need to nil out `State` in this case? I don't think so, but just checking.

**Review by pav-kv — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136156)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @tbg)

___
*[`pkg/kv/kvserver/kvserverpb/proposer_kv.go` line 94 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/136156#-OCnWO_u0HbtXefS1onS:-OCnb-1v2ucPZSRrFT35:bld61ln) ([raw file](https://github.com/cockroachdb/cockroach/blob/873795935fbfe1f5eb3f2d4a1e87038b93d46af4/pkg/kv/kvserver/kvserverpb/proposer_kv.go#L94)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

This is doing more than I expected. Does this method make sense? It's called only once in a specific situation, separating this out makes things less clear if anything.

Also do we have to worry about the case in which `r.State` contains only the `TruncatedState`? Do we need to nil out `State` in this case? I don't think so, but just checking.
</blockquote></details>

Inlined. I don't think it needs to set `r.State` to nil (at least the previous code did not).

**Comment by pav-kv** [comment]
TFTR!

bors r=tbg

**Pull request merged**